shards/core/coro.cpp

```diff
@@ -1,34 +1,132 @@
+#if SH_USE_THREAD_FIBER
+
+ThreadFiber::~ThreadFiber() {}
+
+void ThreadFiber::init(std::function<void()> fn) {
+  isRunning = true;
+
+  // Initial suspend
+  thread.emplace([=]() {
+    fn();
+
+    // Final suspend
+    finished = true;
+    switchToCaller();
+  });
+
+  // Wait for initial suspend
+  std::unique_lock<decltype(mtx)> l(mtx);
+  while (isRunning) {
+    cv.wait(l);
+  }
+}
+
+void ThreadFiber::resume() {
+  if (isRunning) {
+    switchToCaller();
+  } else {
+    // SPDLOG_TRACE("CALLER> {}", thread->get_id());
+    switchToThread();
+    // SPDLOG_TRACE("CALLER< {}", thread->get_id());
+    if (finished) {
+      thread->join();
+      thread.reset();
+    }
+  }
+}
+
+void ThreadFiber::suspend() {
+  // SPDLOG_TRACE("RUNNER< {}", thread->get_id());
+  assert(isRunning);
+  switchToCaller();
+  // SPDLOG_TRACE("RUNNER> {}", thread->get_id());
+}
+
+void ThreadFiber::switchToCaller() {
+  std::unique_lock<decltype(mtx)> l(mtx);
+  isRunning = false;
+  cv.notify_all();
+  if (!finished) {
+    while (!isRunning) {
+      cv.wait(l);
+    }
+  }
+}
+void ThreadFiber::switchToThread() {
+  std::unique_lock<decltype(mtx)> l(mtx);
+  isRunning = true;
+  cv.notify_all();
+  while (isRunning) {
+    cv.wait(l);
+  }
+}
+
+ThreadFiber::operator bool() const { return !finished; }
+
+#else // SH_USE_THREAD_FIBER
+#ifndef __EMSCRIPTEN__
+
+Fiber::Fiber(SHStackAllocator allocator) : allocator(allocator) {}
+void Fiber::init(std::function<void()> fn) {
+  continuation.emplace(boost::context::callcc(std::allocator_arg, allocator, [=](boost::context::continuation &&sink) {
+    continuation.emplace(std::move(sink));
+    fn();
+    return std::move(continuation.value());
+  }));
+}
+void Fiber::resume() {
+  assert(continuation);
+  continuation = continuation->resume();
+}
+void Fiber::suspend() {
+  assert(continuation);
+  continuation = continuation->resume();
+}
+Fiber::operator bool() const { return continuation.has_value() && (bool)continuation.value(); }
+
+#else // __EMSCRIPTEN__
 
 thread_local emscripten_fiber_t *em_local_coro{nullptr};
 thread_local emscripten_fiber_t em_main_coro{};
-thread_local uint8_t em_asyncify_main_stack[SHCoro::as_stack_size];
+thread_local uint8_t em_asyncify_main_stack[Fiber::as_stack_size];
 
 [[noreturn]] static void action(void *p) {
   SHLOG_TRACE("EM FIBER ACTION RUN");
-  auto coro = reinterpret_cast<SHCoro *>(p);
+  auto coro = reinterpret_cast<Fiber *>(p);
   coro->func();
   // If entry_func returns, the entire program will end, as if main had
   // returned.
   abort();
 }
 
-void SHCoro::init(const std::function<void()> &func) {
+void Fiber::init(const std::function<void()> &func) {
   SHLOG_TRACE("EM FIBER INIT");
   this->func = func;
   c_stack = new (std::align_val_t{16}) uint8_t[stack_size];
   emscripten_fiber_init(&em_fiber, action, this, c_stack, stack_size, asyncify_stack, as_stack_size);
+
+  // Initial resume to match functionality of other Fiber implementations
+  resume();
```

Comment:
This might be breaking wasm, behavior changes from past no?

Reply:
Only for emscripten to match the behavior of boost callcc (which resumes initially, emscripten doesn't)
I moved it here, the resume was initially in the prepare call itself